csharp/span_based_serialization.md

```diff
@@ -0,0 +1,101 @@
+
+- Allows parsing directly from both managed and native memory buffer in a safe manner with a single API (this is beneficial for Grpc.Core implementation that uses a native layer, so we don't have to copy buffers into managed memory first).
+
+- In grpc-dotnet, the requests can be exposed as `ReadOnlySequence<byte>` in a very efficient manner.
+
+- Buffer segments can be reused once read so that there are no extra memory allocations.
+
+- Access to segments of read only sequence via `ReadOnlySpan<>` is heavily optimized in recent .NET Core releases.
+
+`WriteTo(ref CodedOutputWriter output)` allows serializing protobuf messages to `IBufferWriter<byte>` which has these benefits:
+
+- The output doesn't have to be written to a single continguous chunk of memory, but into smaller buffer segments, which
+avoids allocating large blocks of memory, resulting in slow LOH GC collections.
+
+- Allows serializing to both managed and native memory in a safe manner with the same API.
+
+- Buffer segments can be reused once read so that there are no extra memory allocations.
+
+- Allows avoiding any additional copies of data when serializing (in both Grpc.Core and grpc-dotnet). 
+
+## Platform support
+
+- the new serialization API will be supported for projects that support at least `netstandard2.0` and allow use of C# 7.2. On newer .NET frameworks, there might be (and likely will be) further performance benefits.
+
+- the new serialization API will be supported on `net45`, but it's mostly for convenience and to minimize the differences between different builds of the library. `netstandard2.0` and higher are the primary targets.
+
+- the new serialization API won't be supported on netstandard1.0 (as it doesn't support the `System.Memory` package which contains several required types). Users will still be able to use the original serialization APIs as always. 
+
+## Backwards compatibility
+
+API changes to the runtime library are purely additive.
+
+By default (without `--use_buffer_serialization` options), the generated code will stay the same and won't depend on any of the newly introduced types, so existing users that don't opt-in won't be affected by the new APIs at all.
+
+When `--use_buffer_serialization` is used when generating the code, the code requiring the newly introduces types will
+be protected by `PROTOBUF_DISABLE_BUFFER_SERIALIZATION` if-define, so that it can be disabled (needs to be manually set in a project) when needed.
+
+## Concern: Code duplication and maintainability
+
+While the logical structure of the (de)serialization code for both old and new approaches is very similar, due to the different nature of the object holding the parsing state (`class` vs `ref struct`), the parsing primitives need to be defined twice (e.g. ReadRawVarint64 exist in two copies, one under CodedInputStream and one user CodedInputReader) and that adds some extra overhead when:
```

Comment:
@haberman according to you, how much of a concern is this? Is there a precedent for this within protobuf (having 2 sets of code for serialization/parsing in one language)?

Reply:
We had this as a transitional state when Gerben was rewriting the parser, but that was just temporary. Java is in a state where there are two very different models for generated code (one table-based), but I'm not sure how much extra runtime come

Looking at the CL, I do worry that this is an awful lot of new code. If I filter out all the generated code, this CL is still +7,000 lines of code, which is worrying.

```
$ git diff --stat master ':!csharp/src/Google.Protobuf.Test.TestProtos' ':!csharp/src/Google.Protobuf.Benchmarks/WrapperBenchmarkMessages.cs' ':!csharp/src/Google.Protobuf/Reflection/Descriptor.cs' ':!csharp/src/Google.Protobuf/WellKnownTypes' ':!csharp/src/Google.Protobuf.Benchmarks/Benchmarks.cs' ':!csharp/src/Google.Protobuf.Benchmarks/BenchmarkMessage1Proto3.cs' ':!csharp/src/AddressBook/Addressbook.cs'
 .gitignore                                                                                 |    3 +
 Makefile.am                                                                                |   15 +-
 conformance/Makefile.am                                                                    |    1 +
 csharp/generate_protos.sh                                                                  |   11 +-
 csharp/src/Google.Protobuf.Benchmarks/Google.Protobuf.Benchmarks.csproj                    |   10 +-
 csharp/src/Google.Protobuf.Benchmarks/GoogleMessageBenchmark.cs                            |  114 ++++++++++
 csharp/src/Google.Protobuf.Benchmarks/Program.cs                                           |    3 +-
 csharp/src/Google.Protobuf.Benchmarks/SerializationBenchmark.cs                            |  118 +++++++++-
 csharp/src/Google.Protobuf.Benchmarks/SerializationConfig.cs                               |   16 +-
 csharp/src/Google.Protobuf.Benchmarks/WrapperBenchmark.cs                                  |   47 +++-
 csharp/src/Google.Protobuf.Conformance/Conformance.cs                                      |  276 ++++++++++++++++++++++-
 csharp/src/Google.Protobuf.Conformance/Google.Protobuf.Conformance.csproj                  |    5 +
 csharp/src/Google.Protobuf.Conformance/Program.cs                                          |   89 ++++++--
 csharp/src/Google.Protobuf.Test/Buffers/ArrayBufferWriter.cs                               |  215 ++++++++++++++++++
 csharp/src/Google.Protobuf.Test/Buffers/MaxSizeHintBufferWriter.cs                         |   76 +++++++
 csharp/src/Google.Protobuf.Test/Buffers/ReadOnlySequenceFactory.cs                         |  130 +++++++++++
 csharp/src/Google.Protobuf.Test/{CodedInputStreamExtensions.cs => CodedInputExtensions.cs} |   18 +-
 csharp/src/Google.Protobuf.Test/CodedInputReaderTest.cs                                    |  419 ++++++++++++++++++++++++++++++++++
 csharp/src/Google.Protobuf.Test/CodedInputStreamTest.cs                                    |  425 ++++++++++------------------------
 csharp/src/Google.Protobuf.Test/CodedInputTestBase.cs                                      |  279 +++++++++++++++++++++++
 csharp/src/Google.Protobuf.Test/CodedOutputStreamTest.cs                                   |  398 ++++++++++++++------------------
 csharp/src/Google.Protobuf.Test/CodedOutputTestBase.cs                                     |  294 ++++++++++++++++++++++++
 csharp/src/Google.Protobuf.Test/CodedOutputWriterTest.cs                                   |  314 ++++++++++++++++++++++++++
 csharp/src/Google.Protobuf.Test/Collections/RepeatedFieldTest.cs                           |   55 ++++-
 csharp/src/Google.Protobuf.Test/ExtensionSetTest.cs                                        |   14 +-
 csharp/src/Google.Protobuf.Test/GeneratedMessageTest.Proto2.cs                             |   20 +-
 csharp/src/Google.Protobuf.Test/GeneratedMessageTest.cs                                    |  206 +++++++++++++----
 csharp/src/Google.Protobuf.Test/Google.Protobuf.Test.csproj                                |    4 +
 csharp/src/Google.Protobuf.Test/JsonParserTest.cs                                          |    4 +
 csharp/src/Google.Protobuf.Test/JsonTokenizerTest.cs                                       |    4 +
 csharp/src/Google.Protobuf.Test/MessageParsingHelpers.cs                                   |  145 ++++++++++++
 csharp/src/Google.Protobuf.Test/UnknownFieldSetTest.cs                                     |   24 +-
 csharp/src/Google.Protobuf.Test/WellKnownTypes/WrappersTest.cs                             |  150 ++++++------
 csharp/src/Google.Protobuf.Test/testprotos.pb                                              |  Bin 330948 -> 327492 bytes
 csharp/src/Google.Protobuf/ByteString.cs                                                   |    7 +-
 csharp/src/Google.Protobuf/CodedInputReader.cs                                             | 1339 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 csharp/src/Google.Protobuf/CodedOutputWriter.cs                                            |  702 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 csharp/src/Google.Protobuf/Collections/MapField.cs                                         |  121 +++++++++-
 csharp/src/Google.Protobuf/Collections/RepeatedField.cs                                    |   87 ++++++-
 csharp/src/Google.Protobuf/ExtensionRegistry.cs                                            |    4 +-
 csharp/src/Google.Protobuf/ExtensionSet.cs                                                 |   48 +++-
 csharp/src/Google.Protobuf/ExtensionValue.cs                                               |   57 ++++-
 csharp/src/Google.Protobuf/FieldCodec.cs                                                   |  480 ++++++++++++++++++++++++++++++++++++---
 csharp/src/Google.Protobuf/Google.Protobuf.csproj                                          |   21 +-
 csharp/src/Google.Protobuf/IBufferMessage.cs                                               |   57 +++++
 csharp/src/Google.Protobuf/IMessage.cs                                                     |    2 +-
 csharp/src/Google.Protobuf/MessageExtensions.cs                                            |   25 ++
 csharp/src/Google.Protobuf/MessageParser.cs                                                |   81 +++++++
 csharp/src/Google.Protobuf/SequenceReader.cs                                               |  417 ++++++++++++++++++++++++++++++++++
 csharp/src/Google.Protobuf/UnknownField.cs                                                 |   53 +++++
 csharp/src/Google.Protobuf/UnknownFieldSet.cs                                              |  105 ++++++++-
 src/google/protobuf/compiler/csharp/csharp_bootstrap_unittest.cc                           |    4 +-
 src/google/protobuf/compiler/csharp/csharp_field_base.cc                                   |    2 +-
 src/google/protobuf/compiler/csharp/csharp_field_base.h                                    |    2 +
 src/google/protobuf/compiler/csharp/csharp_generator.cc                                    |    2 +
 src/google/protobuf/compiler/csharp/csharp_map_field.cc                                    |   13 ++
 src/google/protobuf/compiler/csharp/csharp_map_field.h                                     |    2 +
 src/google/protobuf/compiler/csharp/csharp_message.cc                                      |  159 ++++++++++---
 src/google/protobuf/compiler/csharp/csharp_message.h                                       |    3 +
 src/google/protobuf/compiler/csharp/csharp_message_field.cc                                |   10 +
 src/google/protobuf/compiler/csharp/csharp_message_field.h                                 |    2 +
 src/google/protobuf/compiler/csharp/csharp_options.h                                       |    7 +-
 src/google/protobuf/compiler/csharp/csharp_primitive_field.cc                              |   10 +
 src/google/protobuf/compiler/csharp/csharp_primitive_field.h                               |    2 +
 src/google/protobuf/compiler/csharp/csharp_repeated_enum_field.cc                          |   13 ++
 src/google/protobuf/compiler/csharp/csharp_repeated_enum_field.h                           |    2 +
 src/google/protobuf/compiler/csharp/csharp_repeated_message_field.cc                       |   13 ++
 src/google/protobuf/compiler/csharp/csharp_repeated_message_field.h                        |    2 +
 src/google/protobuf/compiler/csharp/csharp_repeated_primitive_field.cc                     |   13 ++
 src/google/protobuf/compiler/csharp/csharp_repeated_primitive_field.h                      |    2 +
 src/google/protobuf/compiler/csharp/csharp_wrapper_field.cc                                |   32 +++
 src/google/protobuf/compiler/csharp/csharp_wrapper_field.h                                 |    4 +
 72 files changed, 6986 insertions(+), 821 deletions(-)
```

The doubled generated code seems like a potential concern for code size also. I guess C# probably isn't used much in any kind of mobile (phone) deployment, but probably some people eventually have code size concerns?

A few specific questions:
1. Once `Span` is available everywhere, could the old code simply become a client of the new code?
1. Is it possible that some of the code can be shared by using stateless functions where all input and output is represented as explicit parameters?
1. Is it necessary to expose the new reader/writer types to the user? Ideally the user could take advantage of the new code without needing to construct a `CodedInputReader` for example, but could just directly pass the `ReadOnlySequence<byte>` or whatever that they want to parse from.

Reply:
> We had this as a transitional state when Gerben was rewriting the parser, but that was just temporary. Java is in a state where there are two very different models for generated code (one table-based), but I'm not sure how much extra runtime come
> 
> Looking at the CL, I do worry that this is an awful lot of new code. If I filter out all the generated code, this CL is still +7,000 lines of code, which is worrying.

Yes, it's a big PR. Here's some more insights:
- A lot of the code changes is tests - we've added a lot of new tests and where possible, we modified the tests to test both old and the new parser at the same time (so some line changes in the tests are just moving tests around - e.g. from CodedInputStreamTest to CodedInputTestBase where it can be used to ). Many of the changes in the tests are mechanic - just making sure that both serializer/parser implementations get tested

- The changes in the codegen plugin (src/google/protobuf/compiler/csharp) are very mechanic and barely have any logic

- The actual new logic is basically two files - CodedInputReader and CodedOutputWriter ("new" version of CodedInputStream and CodedOuputStream) and adding support for those in things like Maps, RepeatedFields etc.   Altogether this is ~2500 lines but CodedInputReader and CodedOuputWriter is modeled after CodedInputStream and CodedOutputStream and the high-level logic stays the same. It's the low-level read from buffer operations and parsing the primitives that's different.

> 
> ```
> $ git diff --stat master ':!csharp/src/Google.Protobuf.Test.TestProtos' ':!csharp/src/Google.Protobuf.Benchmarks/WrapperBenchmarkMessages.cs' ':!csharp/src/Google.Protobuf/Reflection/Descriptor.cs' ':!csharp/src/Google.Protobuf/WellKnownTypes' ':!csharp/src/Google.Protobuf.Benchmarks/Benchmarks.cs' ':!csharp/src/Google.Protobuf.Benchmarks/BenchmarkMessage1Proto3.cs' ':!csharp/src/AddressBook/Addressbook.cs'
>  .gitignore                                                                                 |    3 +
>  Makefile.am                                                                                |   15 +-
>  conformance/Makefile.am                                                                    |    1 +
>  csharp/generate_protos.sh                                                                  |   11 +-
>  csharp/src/Google.Protobuf.Benchmarks/Google.Protobuf.Benchmarks.csproj                    |   10 +-
>  csharp/src/Google.Protobuf.Benchmarks/GoogleMessageBenchmark.cs                            |  114 ++++++++++
>  csharp/src/Google.Protobuf.Benchmarks/Program.cs                                           |    3 +-
>  csharp/src/Google.Protobuf.Benchmarks/SerializationBenchmark.cs                            |  118 +++++++++-
>  csharp/src/Google.Protobuf.Benchmarks/SerializationConfig.cs                               |   16 +-
>  csharp/src/Google.Protobuf.Benchmarks/WrapperBenchmark.cs                                  |   47 +++-
>  csharp/src/Google.Protobuf.Conformance/Conformance.cs                                      |  276 ++++++++++++++++++++++-
>  csharp/src/Google.Protobuf.Conformance/Google.Protobuf.Conformance.csproj                  |    5 +
>  csharp/src/Google.Protobuf.Conformance/Program.cs                                          |   89 ++++++--
>  csharp/src/Google.Protobuf.Test/Buffers/ArrayBufferWriter.cs                               |  215 ++++++++++++++++++
>  csharp/src/Google.Protobuf.Test/Buffers/MaxSizeHintBufferWriter.cs                         |   76 +++++++
>  csharp/src/Google.Protobuf.Test/Buffers/ReadOnlySequenceFactory.cs                         |  130 +++++++++++
>  csharp/src/Google.Protobuf.Test/{CodedInputStreamExtensions.cs => CodedInputExtensions.cs} |   18 +-
>  csharp/src/Google.Protobuf.Test/CodedInputReaderTest.cs                                    |  419 ++++++++++++++++++++++++++++++++++
>  csharp/src/Google.Protobuf.Test/CodedInputStreamTest.cs                                    |  425 ++++++++++------------------------
>  csharp/src/Google.Protobuf.Test/CodedInputTestBase.cs                                      |  279 +++++++++++++++++++++++
>  csharp/src/Google.Protobuf.Test/CodedOutputStreamTest.cs                                   |  398 ++++++++++++++------------------
>  csharp/src/Google.Protobuf.Test/CodedOutputTestBase.cs                                     |  294 ++++++++++++++++++++++++
>  csharp/src/Google.Protobuf.Test/CodedOutputWriterTest.cs                                   |  314 ++++++++++++++++++++++++++
>  csharp/src/Google.Protobuf.Test/Collections/RepeatedFieldTest.cs                           |   55 ++++-
>  csharp/src/Google.Protobuf.Test/ExtensionSetTest.cs                                        |   14 +-
>  csharp/src/Google.Protobuf.Test/GeneratedMessageTest.Proto2.cs                             |   20 +-
>  csharp/src/Google.Protobuf.Test/GeneratedMessageTest.cs                                    |  206 +++++++++++++----
>  csharp/src/Google.Protobuf.Test/Google.Protobuf.Test.csproj                                |    4 +
>  csharp/src/Google.Protobuf.Test/JsonParserTest.cs                                          |    4 +
>  csharp/src/Google.Protobuf.Test/JsonTokenizerTest.cs                                       |    4 +
>  csharp/src/Google.Protobuf.Test/MessageParsingHelpers.cs                                   |  145 ++++++++++++
>  csharp/src/Google.Protobuf.Test/UnknownFieldSetTest.cs                                     |   24 +-
>  csharp/src/Google.Protobuf.Test/WellKnownTypes/WrappersTest.cs                             |  150 ++++++------
>  csharp/src/Google.Protobuf.Test/testprotos.pb                                              |  Bin 330948 -> 327492 bytes
>  csharp/src/Google.Protobuf/ByteString.cs                                                   |    7 +-
>  csharp/src/Google.Protobuf/CodedInputReader.cs                                             | 1339 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
>  csharp/src/Google.Protobuf/CodedOutputWriter.cs                                            |  702 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++
>  csharp/src/Google.Protobuf/Collections/MapField.cs                                         |  121 +++++++++-
>  csharp/src/Google.Protobuf/Collections/RepeatedField.cs                                    |   87 ++++++-
>  csharp/src/Google.Protobuf/ExtensionRegistry.cs                                            |    4 +-
>  csharp/src/Google.Protobuf/ExtensionSet.cs                                                 |   48 +++-
>  csharp/src/Google.Protobuf/ExtensionValue.cs                                               |   57 ++++-
>  csharp/src/Google.Protobuf/FieldCodec.cs                                                   |  480 ++++++++++++++++++++++++++++++++++++---
>  csharp/src/Google.Protobuf/Google.Protobuf.csproj                                          |   21 +-
>  csharp/src/Google.Protobuf/IBufferMessage.cs                                               |   57 +++++
>  csharp/src/Google.Protobuf/IMessage.cs                                                     |    2 +-
>  csharp/src/Google.Protobuf/MessageExtensions.cs                                            |   25 ++
>  csharp/src/Google.Protobuf/MessageParser.cs                                                |   81 +++++++
>  csharp/src/Google.Protobuf/SequenceReader.cs                                               |  417 ++++++++++++++++++++++++++++++++++
>  csharp/src/Google.Protobuf/UnknownField.cs                                                 |   53 +++++
>  csharp/src/Google.Protobuf/UnknownFieldSet.cs                                              |  105 ++++++++-
>  src/google/protobuf/compiler/csharp/csharp_bootstrap_unittest.cc                           |    4 +-
>  src/google/protobuf/compiler/csharp/csharp_field_base.cc                                   |    2 +-
>  src/google/protobuf/compiler/csharp/csharp_field_base.h                                    |    2 +
>  src/google/protobuf/compiler/csharp/csharp_generator.cc                                    |    2 +
>  src/google/protobuf/compiler/csharp/csharp_map_field.cc                                    |   13 ++
>  src/google/protobuf/compiler/csharp/csharp_map_field.h                                     |    2 +
>  src/google/protobuf/compiler/csharp/csharp_message.cc                                      |  159 ++++++++++---
>  src/google/protobuf/compiler/csharp/csharp_message.h                                       |    3 +
>  src/google/protobuf/compiler/csharp/csharp_message_field.cc                                |   10 +
>  src/google/protobuf/compiler/csharp/csharp_message_field.h                                 |    2 +
>  src/google/protobuf/compiler/csharp/csharp_options.h                                       |    7 +-
>  src/google/protobuf/compiler/csharp/csharp_primitive_field.cc                              |   10 +
>  src/google/protobuf/compiler/csharp/csharp_primitive_field.h                               |    2 +
>  src/google/protobuf/compiler/csharp/csharp_repeated_enum_field.cc                          |   13 ++
>  src/google/protobuf/compiler/csharp/csharp_repeated_enum_field.h                           |    2 +
>  src/google/protobuf/compiler/csharp/csharp_repeated_message_field.cc                       |   13 ++
>  src/google/protobuf/compiler/csharp/csharp_repeated_message_field.h                        |    2 +
>  src/google/protobuf/compiler/csharp/csharp_repeated_primitive_field.cc                     |   13 ++
>  src/google/protobuf/compiler/csharp/csharp_repeated_primitive_field.h                      |    2 +
>  src/google/protobuf/compiler/csharp/csharp_wrapper_field.cc                                |   32 +++
>  src/google/protobuf/compiler/csharp/csharp_wrapper_field.h                                 |    4 +
>  72 files changed, 6986 insertions(+), 821 deletions(-)
> ```
> 
> The doubled generated code seems like a potential concern for code size also. I guess C# probably isn't used much in any kind of mobile (phone) deployment, but probably some people eventually have code size concerns?
> 
> A few specific questions:
> 
> 1. Once `Span` is available everywhere, could the old code simply become a client of the new code?

The problem is that Span<> is a `ref struct` (a struct that is always passed as a reference - a new concept in C#) and as such it can only live on the stack (which also acts safety measure if you're e.g. accessing native memory). So you cannot have regular classes such CodedInputStream hold an instance of Span  - that makes it difficult to retrofit the "old" parser to use the CodedInputReader logic internally (currently not possible without degrading the parser's performance).
At some point in the distant future, we could recommend everyone moving to CodedInputReader and CodedOuputWriter, make CodedInputStream and CodedOuputStream "legacy" and make them use the new parsing logic at the expense of worse performance, but still maintaining API compatibility.

> 2. Is it possible that some of the code can be shared by using stateless functions where all input and output is represented as explicit parameters?

We've spent a lot of time investigating this and did as much de-duplication as seemed practical.

> 3. Is it necessary to expose the new reader/writer types to the user? Ideally the user could take advantage of the new code without needing to construct a `CodedInputReader` for example, but could just directly pass the `ReadOnlySequence<byte>` or whatever that they want to parse from.

The CodedInputReader and CodedOutputWriter types need to be public as they are referenced by the generated code (and they hold the intermediate parsing/serialization state - same idea as CodedInputStream/CodedOuputStream). 
Most users won't use these types directly, and would instead call e.g. SomeMessage.Parser.ParseFrom(ReadOnlySequence<byte>)